src/Discovery/Walk.hs

```diff
@@ -148,7 +167,11 @@ walkDir handler topdir =
         Nothing -> pure $ Just ()
         Just traversed' -> walktree traversed' curdir
     walktree traversed curdir = do
-      (subdirs, files) <- listDir curdir
+      -- If we can't list the directory contents, we just skip the directory
+      -- and claim that it has no children.  We don't log it because it would
+      -- happen for every discovery process and would be very noisy.
+      -- TODO: use a (State (Set SkippedDirs)) effect to track and report this
+      (subdirs, files) <- fmap (fromMaybe ([], [])) $ recover $ listDir curdir
```

Comment:
Any reason not to use `fromMaybe ([], []) <$> recover $ listDir curdir`?  It doesn't seem to be a firm choice in our style guide, but I like that it makes the read order a bit more consistent instead of having to jump way over to the `fmap` to see how that affects the `fromMaybe`.

Reply:
I think it makes it easier to read since there's no operator confusion this way, but with `a <$> b $ c` it can sometimes not be clear where the functor operation reaches to.  This particular case is a bad example, but this pattern chains well.

Reply:
Also, I think that functions like `fmap` aren't really interesting in terms of reviewing.  If the types are strict, and there's only one way a function could work with the right type, I just generally assume that it's laid out right.  If we had to check that kind of thing anyway, what's the point of having the strict types?

Reply:
I don't understand what you mean about "aren't really interesting in terms of reviewing".

I'm just pointing out that as a reader this line is a little awkward because it mixes prefix and infix operators. I think leaning more towards parentheses or more towards infix operators would make it easier to read and understand at a glance.

Reply:
Actually, looking closer at this, there's a good argument for making this pattern a diagnostic helper.

Reply:
Sorry, I mispoke.  When said we were mixing operators I meant mixing infix and prefix functions.  It's not really important for my point.

I'm trying to focus on what is easiest to read and understand at a glance.  I found this line harder to read than most and I think we could do something to make it easier to read.  That could be using conventions like `<$>` (which I may just be personally comfortable with), it may be using multiple lines, it may be refactoring the pattern into a helper (as you suggested).  All I'm suggesting is that you could probably make this easier to read before merging, but I'm not blocking on it.

Though I'm confused on one point. When you say "interesting to a reviewer", do you mean me as the code reviewer, or do you mean a general reader reviewing the code?

Reply:
reviewer ~ code reviewer

Also, I made it easier by adding helpers to diagnostics.